src/multitaper.jl

```diff
@@ -0,0 +1,583 @@
+
+Creates a config object which holds the configuration state
+and temporary variables used in multitaper computations,
+e.g. [`mt_pgram!`](@ref), [`mt_spectrogram`](@ref), and [`mt_cross_spectral!`](@ref).
+
+An `MTConfig` can be re-used between computations as long
+as none of the input arguments change.
+
+* `n_samples`: the number of samples to be used as input when computing multitaper periodograms
+  with this configuration. Used for pre-allocating temporary buffers.
+* `fs`: the number of samples per second of the input signal
+* `nfft`: length of input vector to the FFT; if `nfft > n_samples`, then the
+  input signal will be zero-padded until it is of length `nfft`.
+* `window`: window function to use for tapering. If left at the default of `nothing`,
+  `window` will be set to `dpss(n_samples, nw, ntapers)`.
+* `ntapers`: the number of tapers to use.
+* `onesided`: whether or not to compute a "one-sided" FFT by using that real signal data
+  yields conjugate-symmetry in Fourier space.
+* `fft_flags`: flags to control how the FFT plan is generated.
+"""
+function MTConfig{T}(n_samples; fs=1, nfft=nextpow(2, n_samples), window=nothing, nw=4,
+                     ntapers=2 * nw - 1, onesided::Bool=T <: Real,
+                     fft_flags=FFTW.MEASURE) where {T}
+    if onesided && T <: Complex
+        throw(ArgumentError("cannot compute one-sided FFT of a complex signal"))
+    end
+    n_samples > 0 || throw(ArgumentError("`n_samples` must be positive"))
+    nfft >= n_samples || throw(ArgumentError("Must have `nfft >= n_samples`"))
+    ntapers > 0 || throw(ArgumentError("`ntapers` must be positive"))
+    freq = onesided ? rfftfreq(nfft, fs) : fftfreq(nfft, fs)
+    fft_input_tmp = Vector{T}(undef, nfft)
+    fft_output_tmp = Vector{fftouttype(T)}(undef, length(freq))
+    plan = onesided ? plan_rfft(fft_input_tmp; flags=fft_flags) :
+           plan_fft(fft_input_tmp; flags=fft_flags)
+    if window === nothing
+        r = fs * ntapers
+        window = dpss(n_samples, nw, ntapers)
+    else
+        r = fs * sum(abs2, window)
+    end
+
+    return MTConfig{T}(n_samples, nfft, ntapers, freq, fs, plan, fft_input_tmp,
+                       fft_output_tmp, window, onesided, r)
+end
+
+function allocate_output(config::MTConfig{T}) where {T}
+    return Vector{fftabs2type(T)}(undef, length(config.freq))
+end
+
+
+# Internal function used in [`mt_pgram!`](@ref) and [`mt_cross_spectral!`](@ref).
+function mt_fft_tapered!(fft_output, signal, taper, config)
+    # Create the input: tapered + zero-padded version of the signal
+    fft_input = config.fft_input_tmp
+    @inbounds for i in 1:length(signal)
+        fft_input[i] = config.window[i, taper] * signal[i]
+    end
+    fft_input[(length(signal) + 1):(config.nfft)] .= 0
+
+    # do the FFT
+    mul!(fft_output, config.plan, fft_input)
+    return nothing
+end
+
+
+"""
+    mt_pgram!(output::AbstractVector, signal::AbstractVector, config::MTConfig) -> Periodogram
```

Comment:
I think there are four possible methods:

* out of place, pass a config
* in place, pass a config (*)
* out of place, construct the config inside the function (*)
* in place, construct the config inside the function

where the (*) methods are currently implemented (except `mt_spectrogram`, which has a variant on "out of place, pass a config": it has an out of place method where you pass a `MTConfig` instead of a `MTSpectrogramConfig`).

I chose these two as extremes of max-convienence vs max-perf. I can add "in place, construct the config inside the function" as you suggest too, or even both the other options. I was a bit worried about it being too confusing if there are many methods but perhaps it's nice to have all the options.

Reply:
I added the remaining methods so we have all 4 for each function in the latest commit